pilot/pkg/proxy/envoy/v1/watcher.go

```diff
@@ -55,34 +56,48 @@ type CertSource struct {
 }
 
 type watcher struct {
-	agent    proxy.Agent
-	role     model.Node
-	config   meshconfig.ProxyConfig
-	certs    []CertSource
-	pilotSAN []string
+	agent         proxy.Agent
+	role          model.Node
+	config        meshconfig.ProxyConfig
+	requiredCerts []CertSource // These certs must present for Envoy to start
+	certs         []CertSource
+	pilotSAN      []string
 }
 
 // NewWatcher creates a new watcher instance from a proxy agent and a set of monitored certificate paths
 // (directories with files in them)
 func NewWatcher(config meshconfig.ProxyConfig, agent proxy.Agent, role model.Node,
-	certs []CertSource, pilotSAN []string) Watcher {
+	internalCerts, externalCerts []CertSource, pilotSAN []string) Watcher {
 	return &watcher{
-		agent:    agent,
-		role:     role,
-		config:   config,
-		certs:    certs,
-		pilotSAN: pilotSAN,
+		agent:         agent,
+		role:          role,
+		config:        config,
+		requiredCerts: internalCerts,
+		certs:         append(internalCerts, externalCerts...),
+		pilotSAN:      pilotSAN,
 	}
 }
 
 const (
+	requiredCertsCheckInterval = 4 * time.Second
+	requiredCertsCheckTimeout  = 90 * time.Second
 	// defaultMinDelay is the minimum amount of time between delivery of two successive events via updateFunc.
 	defaultMinDelay = 10 * time.Second
 	azRetryInterval = time.Second * 30
 	azRetryAttempts = 10
 )
 
 func (w *watcher) Run(ctx context.Context) {
+	// Wait until the required certificate files present. This prevents Envoy failing caused by non-existing cert files.
+	// Note: this function is blocking.
+	log.Infof("Check required cert files are present (up to %v)...", requiredCertsCheckTimeout)
+	if err := waitForCertsPresent(w.requiredCerts, requiredCertsCheckInterval, requiredCertsCheckTimeout); err != nil {
+		log.Errorf("Required cert file check failed: %v", err)
+		<-ctx.Done()
```

Comment:
Shouldn't this be ctx.Err()

@kyessenov can you help with the review?

Reply:
`<-ctx.Done()`/`ctx.Err()` don't belong here unless the intent is to wait for the caller to cancel the context if the cert watch failed.

Reply:
I see. Thanks @ayj